### ISSUE TITLE
Support pre-post hooks for build, deploy and sync

### ISSUE THREAD

**Issue by adrienjt** [GoogleContainerTools/skaffold#1441]
Hi, in my workflow, I need to run some commands before I run skaffold. In CI/CD, I wrapped skaffold in a script, but it's a pain for developers to remember to run a script before running `skaffold run` ~or `skaffold dev`~ or for me to maintain a wrapping CLI (EDIT: and `skaffold dev` doesn't work, see use case below). The developer experience would be improved if skaffold supported pre-build/pre-deploy hooks (and post-build/post-deploy hooks).

Suggestion:

```yaml
apiVersion: skaffold/v1...
kind: Config
hooks:
  preBuild:
    exec:
      command: ["python", "my-pre-build-hook.py"]
  postBuild:
    exec:
      command: ["./my-post-build-hook.sh"]
  preDeploy:
    exec:
      command: ["terraform", ...]
  postDeploy:
    exec:
      command: ["curl", ...]
...
```

Details on my use case:
- I have a Python monorepo of microservices and shared libraries. I don't want to use the root of the repo as a microservice's Dockerfile context, because of the performance hit (breaks caching, huge context to send to docker daemon or kaniko), so before I build a microservice image, I copy the required shared libraries temporarily in the microservice subfolder / Dockerfile context, and remove them afterward (Docker doesn't follow symlinks if you wonder).
- kustomize doesn't support build-time variables, so I generate overlays to label, prefix and change namespaces depending on, e.g., the pull request number or the developer name.

**Comment by davidkarim** [comment]
Would love to hear other comments regarding whether this type of feature would be desirable. In my case, I do have to add a few steps prior to running skaffold dev (such as retrieving secrets from Vault). I would definitely prefer having a way to add these pre-build hooks in the skaffold config file rather than having to build a separate wrapper script or asking developers to run a shell script prior to running skaffold dev.

**Comment by nthypes** [comment]
@dgageot we are open to ship this?

**Comment by tejal29** [comment]
@adrienjt Is it correct to say the script you have that copies the relevant/required libs to the micro-services folder is actually a way for maintaining python  dependency? Or does it do something more than that?

What would a pre-deploy hook ?

@davidkarim : The steps you mentioned, like retrieving secrets from Vault,  right now we expect them to be one time dev setup.

**Comment by balopat** [comment]
I'm worried to make skaffold into a generic workflow execution engine and supporting the issues that come with that. Skaffold is meant to have a fast, opinionated local workflow and a building block for CI/CD use cases. Having completely arbitrary script calls starts to muddle things. 

Also, the use cases mentioned here are interesting: I think I would very much so prefer helping explicitly with 
- secret management
- docker context composition and
- per user / per PR manifest rendering  
over leaving it to the users to implement it themselves in scripts.

**Comment by nthypes** [comment]
> I'm worried to make skaffold into a generic workflow execution engine and supporting the issues that come with that. Skaffold is meant to have a fast, opinionated local workflow and a building block for CI/CD use cases. Having completely arbitrary script calls starts to muddle things.
> 
> Also, the use cases mentioned here are interesting: I think I would very much so prefer helping explicitly with
> 
> * secret management
> * docker context composition and
> * per user / per PR manifest rendering
> 
> over leaving it to the users to implement it themselves in scripts.

@balopat I understand your fear and I agree with you in parts. 

The point is that this *_ local workflow_* that you mentioned, nowadays is becoming more complex with a lot of tools that developers use during the development workflow. So, having this kind of **_hooks_** help us to keep skaffold as the main entry point for all the local build/test/deployment workflow.

 For example, in our case, we have some _steps/scripts_ that we should execute before the docker build. Instead of using only the file watcher of skaffold, we are using the individual `--watch` flag of those scripts, because we can't control this script execution in skaffold before the build.

**Comment by giladsh1** [comment]
@balopat @tejal29 @nthypes 
I've closed  #2425 in favor of this issue, but I'm copying some of the text I wrote there, in order to emphasize the need for this functionality.

IMHO These callbacks can have major befits for the development / deployment process, and make skaffold a true "one stop shop" tool.

#### some examples -

1. creating a beforeDev callback, which will allow the user to copy some files from a specific location before running skaffold dev. This will allow the user to copy files which are required for building the artifact, but exists in another location and can't be copied into a docker image, since they're not in the docker daemon context.

1. creating a afterDevDeployment callback, which will allow the user to run a custom command after the deployment is done when running the skaffold dev command. This action can help in many different ways, such as to create mock data for dev environments, triggers UI automation tests, and many other things. some things are not easily implemented in a docker image running somewhere.

#### why not use a wrapper
One could argue that these actions can be achieved with a simple wrapper before / after running skaffold. However, some actions (such as dev or debug) run in continuous mode, and can't be easily moved to the background.

**Comment by adrienjt** [comment]
@tejal29 correct, it's a way to ensure internal Python dependencies (in the same repo but outside of the Docker build context) are present in the Docker build context to be copied to the image.

I can think of two use cases for a pre-deploy hook:

1) generate manifests with anything but Helm and kustomize, e.g., Jsonnet (not currently supported by skaffold) or a custom tool (will never be supported by skaffold);
2) deploy non-Kubernetes dependencies, e.g., cloud databases, static assets, etc., with terraform, aws-cli, etc.

@balopat, I agree with @nthypes, @giladsh1 and @davidkarim that skaffold really shines when used directly rather than within a wrapper script, because of its clean developer API (something that's hard to design and maintain) and all the watch ~magic~ logic. With hooks, you'd make skaffold more usable. Therefore, more people would use it. My developer workflow is *mostly* supported by skaffold, but that subtle difference makes implementing skaffold a lot more complicated. Seeing what people would typically do with hooks, you could later implement "cleaner" equivalent features at your own pace.

**Comment by tbarrella** [comment]
Agreeing with what people have said above,
- Docker context composition before builds: Explicit skaffold.yaml configuration for this would be helpful, although supporting custom scripts to e.g. generate files would also be helpful to support migrations. Ideally multi-stage builds would prevent the need for generating files for the context, but not all builders have good support for multi-stage builds, such as Kaniko as of now
- Generating manifests before deploys: An example of this being helpful is when generating large numbers of variants with different parameters using a conventional programming language (see https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/declarative-application-management.md#what-about-configurations-with-high-cyclomatic-complexity-or-massive-numbers-of-variants)

**Comment by AndiDog** [comment]
For https://github.com/GoogleContainerTools/skaffold/issues/1441
::

From the above comments, there seems to be a common requirement of **generating manifests before deploy**. I'd like to add another use case example and proposed solution:

In a GitOps scenario, the CI builds the image(s) through `skaffold build`, generates a manifest, fills in the built image tags and commits the full manifest to some repo. The CD tool would then pick up the committed manifest and deploy it. In such case, Helm or other tools are not involved anymore, but only the full manifest is applied using kubectl / Kubernetes API (at least in case of Argo CD). Let's assume my application uses a Helm template and a hardcoded version of Helm to create the manifest, and then appends per-environment stuff to it (e.g. sealed secrets). Currently, Skaffold would not support this model, but only has the Helm deployer which internally uses `helm` CLI and therefore already differs from what the CD tool would do. For consistency, Skaffold should support deploying in the same fashion.

I think that instead of **hooks**, Skaffold could also solve this particular requirement by providing a **deployer type which runs a custom script to generate the manifest**, fills in image tags and deploys using `kubectl`. Essentially, it can be seen as **small extension to KubectlDeployer** which allows generating manifest files from a script instead of specifying globs to existing files. The "API" could involve setting an environment variable for the chosen Skaffold profile, so the script could distinguish environments. The above comments also contain some more use cases that are possible with this approach (like pulling secrets).

---

Would that be a feature you'd like to see implemented? Specifically, extend KubectlDeployer like:

```
deploy:
  kubectl:
    manifests:
      - /tmp/generated.yml
    create-manifests-script: generate-manifest.sh
```

Script example:

```
#!/bin/sh
# generate-manifest.sh
set -e
helm-at-specific-version template [...] >/tmp/generated.yml
cat k8s/${SKAFFOLD_PROFILE}/per-env-sealed-secrets.yml >>/tmp/generated.yml
```

---

Slightly off-topic, but handy addition: add a Skaffold subcommand which fills in image tags into a manifest from the output of the build stage. The CI steps for the GitOps model would then be `skaffold build --file-output=x`, `skaffold manifest --from-build-output=x`, commit full manifest to repo which CD tool picks up. And developers would consistently use the single source of truth (`skaffold.yaml`) to generate/deploy changes (via `skaffold {dev,debug,run}`).

**Comment by Multiply** [comment]
I might just be bad at reading the docs, but having the ability to run a command inside a container after file-sync would also be really amazing.

**Comment by bitsofinfo** [comment]
I definitely have the need to run multiple custom commands prior to build, as well as after deploy. Such a feature is greatly needed. Right now I'm having to run a wrapper shell script around `skaffold run` to achieve all this.

**Comment by nkubala** [comment]
quick update from the team here: this is something we're actively prioritizing and working on a preliminary design for. no commitment on timeline for release at the moment, but we'll keep everyone posted and ship this as soon as it's ready :)

**Comment by ItsLeeOwen** [comment]
the use case I'm looking for is in regards to secret management, initially for local developer environment to avoid using a script or makefile in front of `skaffold dev` to run a command such as `kubectl create secret generic base --from-env-file=.env`

**Comment by legopin** [comment]
> the use case I'm looking for is in regards to secret management, initially for local developer environment to avoid using a script or makefile in front of `skaffold dev` to run a command such as `kubectl create secret generic base --from-env-file=.env`

There is a different tool supported by skaffold you can use to solve this use case -- `kustomize`'s secret generator.
It allows for secrets to be generated from local files, and allows you to customize local vs production variants in deployment files.

https://kubernetes-sigs.github.io/kustomize/guides/plugins/builtins/#_secretgenerator_

**Comment by blaggacao** [comment]
Ahm... https://github.com/kubernetes/kubernetes/pull/70875#issuecomment-453369450 ?

If skaffold should ever be getting used on such popular repo, how could people be alerted that it's _actually_ a [task runner](https://github.com/casey/just) - for if to avoid `make` is of concern?

**Comment by blaggacao** [comment]
> I wrapped skaffold in a script, but it's a pain for developers to remember to run a script before running skaffold run

Since I'm arguing against arbitrary task runs to be _included_ in skaffold for security dynamics ...

I'd suggest using [`just`](https://github.com/casey/just) as a task runner. It pretends to be one, and nobody would be tricked into thinking it does anything different but execute arbitrary commands.

Moreover, I very clearly hear the motivation to nudge developers into using the correct cli API of one's repository. But a repositories CLI API usually doesn't stop with `skaffold`. Most times, you also have `lint`, `test`, etc.

So increasing the flight-height on this issue, `skaffold` will hopefully elicit the temptation of wanting to be a generic task runner. Rather, it would document suggestions clearly for this use case.

And if one wants to get exciting and auto-execute tasks on file changes, Id recommend: https://github.com/direnv/direnv

And if one wants to get even more exciting in nudging the developer hords into doing the right thing: I'm using nix-shell, specifically https://github.com/xzfc/cached-nix-shell to provide fully customized shell environments with all necessary tools automatically available and updated to my will on the developer's finger tip.

**Comment by nkubala** [comment]
thanks for bringing this up @blaggacao. in some ways this is a necessary evil of providing an "escape hatch" for users to break out of the predefined pipeline stages in skaffold, since we can't really know what they're trying to do beforehand. unfortunately I don't think it's feasible for us to "invert" the control flow in skaffold to have another task runner ala direnv actually execute the pipeline stages of skaffold, this would be a complete paradigm shift and sort of defeat the purpose of having `skaffold dev`.

I will say that I think it makes sense for us to have some sort of safeguard against arbitrarily running a hook script on your local machine, especially in the context of grabbing another pre-existing skaffold.yaml in a project and running `skaffold dev` (which we'd like to make a more common workflow in the future). after talking with the team we're thinking it might make sense to hide hook execution behind a flag, `--enable-hooks`, which would be `false` by default - this would make the user have to be intentional about executing the hooks defined in the skaffold.yaml. we'd also want to make the messaging more clear from skaffold when hooks are actually being executed, so that users know that things could potentially be happening on their machine that they might not be aware of.

**Comment by blaggacao** [comment]
> `--enable-hooks`, which would be false by default

Excellent! - It might make perfectly sense to call out to `just` from `skaffold` hooks and vice-versa, then. Mix and match! Great!

Might you (or is there) include(d) a way to signal to `skaffold dev` a dependency invalidation to trigger a rebuild - just in case I'd go down the route and leverage the arbitrarity-powers of `direnv` as a watcher? (eg. to - hypothetically - do jsonnet pre-rendering?)

You might also opt for a workflow like direnv which you need to opt-in (`direnv allow`). The opt-in is only valid for one specific hash of the `.envrc` file. Once that changes, it tells you to review and `direnv allow` again.

A pretty neat workflow, which you might consider statefully opting in over a hash of all command specs as well for skaffold (giving a little credits to `direnv`, of course :wink: ).

**Comment by nkubala** [comment]
> Might you (or is there) include(d) a way to signal to skaffold dev a dependency invalidation to trigger a rebuild - just in case I'd go down the route and leverage the arbitrarity-powers of direnv as a watcher? (eg. to - hypothetically - do jsonnet pre-rendering?)

your best bet here would be to try and use the control API, which allows you to interact with the dev loop and control individual build/sync/deploy actions manually. you can do this through REST or through a gRPC client: https://skaffold.dev/docs/design/api/

so you could have a little client that listens for `direnv` events, and then also has an outgoing client to your `skaffold dev` process that sends triggers to it through the API when it receives the go-ahead from `direnv`. this is basically a poor man's custom watcher :)

**Comment by carlskii** [comment]
This is exactly what I need to add my CLI that can be used to inject Keys/Certificates into some applications. Would it possible to implement a way to call a function from a custom go application here too?

**Comment by gsquared94** [comment]
@carlskii, you should be able to run your go binaries or any executable. Skaffold doesn't provide any special runtime so it just runs on your machine and/or the deployed containers.
The design doc is [here](https://github.com/GoogleContainerTools/skaffold/blob/master/docs/design_proposals/lifecycle-hooks.md)

**Comment by justinfx** [comment]
Any movement on this, beyond the hooks design doc?

**Comment by tejal29** [comment]
@justinfx not yet. We are looking into it!

**Comment by coopbri** [comment]
I would love this! In the meantime, I've wrapped Skaffold in a script as OP and others have mentioned. I haven't had issues with syncing, but maybe my use cases don't warrant that issue.

Something simple like this:
```sh
#!/bin/sh

# setup
# e.g. make directories, copy files

# run Skaffold
skaffold dev

# teardown
# e.g. remove files
```

**Comment by chermed** [comment]
Another use case is when we use a package manager to manage the dependencies of a project, before building the image we need to generate the locked file to use 

Example using python  and poetry: launch` poetry export -f requirements.txt` before each build

**Comment by actraiser** [comment]
My use case is the generation of dynamic Typescript types for GraphQL schemas. The process of creating this types.d.ts is simply running a cli command on the host. The command parses all schema files and the typescript definition file is being generated on the host and with that synced to the docker container - which is not half bad.

But ideally that cli command to generate the types.d.ts file would be run by skaffold in the projects folder when it detects a change in one of the schema files (identified by some glob). Is this possible with the current feature set of  skaffold? 

Greets
-act

**Comment by gsquared94** [comment]
Thanks all for bearing with us while this issue kept getting reprioritized. We'll start working on this from next month. 
Linking the [design](https://tinyurl.com/skaffold-hooks) that was reviewed a while ago but open to any new feedback.

**Comment by gsquared94** [comment]
This is completed and available in release [v1.31.0](https://github.com/GoogleContainerTools/skaffold/releases/tag/v1.31.0)
